subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskExecution.java

```diff
@@ -345,18 +345,19 @@ public void visitOutputs(File workspace, OutputVisitor visitor) {
     }
 
     @Override
-    public void handleUnreadableInputs(InputFingerprinter.InputFileFingerprintingException ex) {
-        nagUserAboutUnreadableInputsOrOutputs("input", ex.getPropertyName(), ex.getCause());
+    public RuntimeException decorateInputFileFingerprintingException(InputFingerprinter.InputFileFingerprintingException ex) {
```

Comment:
Seems confusing to me that in method signature we return `RuntimeException` but in fact we throw it. I would expect from that method that exception is just decorated and decorated exception is returned.

What if we rather keep void return type and name method something like  `decorateAndThrowFingerprintingException`?

Reply:
The reason I'm returning an exception here is so that I can do `throw decorateInputFileFingerprintingException(...)`, which tells the compiler that execution will not flow past that call. I cannot do that with a `void`-returning method. I stole this pattern from Guava's [`Throwables.propagate()`](https://guava.dev/releases/30.0-jre/api/docs/com/google/common/base/Throwables.html#propagate(java.lang.Throwable)).

> This method always throws an exception. The `RuntimeException` return type allows
> client code to signal to the compiler that statements after the call are unreachable. Example
> usage:
> 
> ```
> T doSomething() {
>   try {
>     return someMethodThatCouldThrowAnything();
>   } catch (IKnowWhatToDoWithThisException e) {
>     return handle(e);
>   } catch (Throwable t) {
>     throw Throwables.propagate(t);
>   }
> }
> ```